packages/@react-aria/grid/src/useGrid.ts

```diff
@@ -114,36 +120,40 @@ export function useGrid<T>(props: GridProps, state: GridState<T, GridCollection<
   let selection = state.selectionManager.rawSelection;
   let lastSelection = useRef(selection);
   useUpdateEffect(() => {
-    // Do not do this when using selectionBehavior = 'replace' to avoid selection announcements
-    // every time the user presses the arrow keys.
-    if (!state.selectionManager.isFocused || state.selectionManager.selectionBehavior === 'replace') {
-      lastSelection.current = selection;
-
+    if (!state.selectionManager.isFocused) {
       return;
     }
 
     let addedKeys = diffSelection(selection, lastSelection.current);
     let removedKeys = diffSelection(lastSelection.current, selection);
 
     // If adding or removing a single row from the selection, announce the name of that item.
+    let isReplace = state.selectionManager.selectionBehavior === 'replace';
     let messages = [];
-    if (addedKeys.size === 1 && removedKeys.size === 0) {
+    if (addedKeys.size === 1 && (removedKeys.size === 0 || (state.selectionManager.selectedKeys.size === 1 && isReplace))) {
       let addedText = getRowText(addedKeys.keys().next().value);
       if (addedText) {
         messages.push(formatMessage('selectedItem', {item: addedText}));
       }
-    } else if (removedKeys.size === 1 && addedKeys.size === 0) {
+    } else if (removedKeys.size === 1 && (addedKeys.size === 0 || (state.selectionManager.selectedKeys.size === 1 && isReplace))) {
       if (state.collection.getItem(removedKeys.keys().next().value)) {
         let removedText = getRowText(removedKeys.keys().next().value);
         if (removedText) {
           messages.push(formatMessage('deselectedItem', {item: removedText}));
         }
       }
+    } else if (isReplace && (lastSelection.current === 'all' || removedKeys.size > 0) && state.selectionManager.selectedKeys.size === 1) {
```

Comment:
does it matter if keys were removed, or should we just always announce the first selected item if there is one? If you moved this branch above the others, you wouldn't need the added or condition either?